### PR TITLE
feat(layouts): Flow container enhancements

### PR DESCRIPTION
**Pull request by wodenx**
## Changes
- Fix crash when using `startWith`
- Fix issues when components have no titles or are missing mandatory categories.
- Add `blacklistCategories` prop to prevent some metadata from appearing in component selector filters
- Add `withAllTitlesFromTerms` helper which will assign default titles to all components based on their terms
- Improve documentation.

## Test Instructions
- Light regression of flow contaienr page
- New fc-basics page exercises the new funcitonality

### CODE REVIEW

File: examples/test-site/src/data/pages/fc-basics/Basic.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// Create a function which takes a set of tailwind classes and returns
+// a token which constrains the widths of flow container items. This
+// is usually done at the site level, using the sites tailwind config.
+export const withWidthConstraints = flow(
+  resolveConfig,
+  withTailwindWidthConstraints,
+)(tailwindConfig);
+
+const asBasicFlowContainer = asToken(
+  // withDesign(basicDesign),
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/Basic.tsx:91]: Actually - this commented code is intentional - this is a demo page which is meant to illustrate different options for configuring flow containers, so it's intended for a developer to see.

File: packages/bodiless-layouts/src/ComponentSelector/ItemList.tsx [hunk 2]
```diff
@@ -33,6 +33,13 @@ const ItemList: React.FC<ItemListProps> = props => {
   const { components, onSelect } = props;
   const finalUI = useContext(uiContext);
   const [scale, setScale] = useState(Scale.Full);
+
+  // Function to build a default title for a component from its categories.
+  const title = (component: ComponentWithMeta) => component.title || component.displayName;
+
+  // Function to build a defaul t description for a component from its categories.
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/src/ComponentSelector/ItemList.tsx:40]:
```suggestion
  // Function to build a default description for a component from its categories.
```

File: packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      ],
+      [] as string[],
+    )
+    .filter(t => !blacklistTerms.includes(t))
+    .reverse()
+    .join(separator);
+};
+
+/**
+ * Createw a token which adds a `title` property to a component. This is derived from all
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx:56]:
```suggestion
 * Creates a token which adds a `title` property to a component. This is derived from all
```

File: packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * terms which have been applied to that component.
+ *
+ * The new title will only be added if the component does not aleady have a title.
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx:59]:
```suggestion
 * The new title will only be added if the component does not already have a title.
```

File: packages/fclasses/src/Design.tsx [hunk 5]
```diff
@@ -412,9 +412,41 @@ type DesignOrHod<C extends DesignableComponents> = Design<C> | HOD<C>;
...
 export const varyDesign = flowDesignsWith(varyDesign$);
+
+/**
+ * @deprecated
+ * @private
+ */
 export const extendDesign = flowDesignsWith(extendDesign$);
+
+/**
+ * Creates a new desigh which constists of all possible combindations of the
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/fclasses/src/Design.tsx:434]:
```suggestion
 * Creates a new design which consists of all possible combinations of the
```

File: examples/test-site/src/data/pages/fc-basics/index.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright © 2019 Johnson & Johnson
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/index.tsx:2]:
```suggestion
 * Copyright © 2021 Johnson & Johnson
```

File: examples/test-site/src/data/pages/fc-basics/Basic.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    pick(bgColors, 'Blue'),
+    pick(borderColors, 'Teal'),
+  ),
+  ...varyDesigns(
+    pick(bgColors, 'Teal'),
+    pick(borderColors, 'Blue'),
+  ),
+};
+
+// We define a custom preview.  These tokens will only be applied
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/Basic.tsx:59]: I think 2 spaces after a period is an acceptable style.

File: packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ */
+const withTitleFromTerms = (ops: WithTitleFromTermsProps = {}): Token => (
+  (Component: ComponentOrTag<any>) => asToken({
+    title: getTitleFromTerms(Component as ComponentWithMeta<any>, ops),
+  })(Component)
+  // (Component: ComponentOrTag<any>) => {
+  //   const WithTitleFromTerms: ComponentWithMeta = (props: any) => <Component {...props} />;
+  //   WithTitleFromTerms.title = getTitleFromTerms(Component as ComponentWithMeta<any>, ops);
+  //   return WithTitleFromTerms;
+  // }
```
Comment (wodenx) [inline packages/bodiless-layouts/src/meta/withTitleFromTerms.tsx:75]:
```suggestion
```

File: packages/bodiless-layouts/README.md [hunk 9]
```diff
@@ -70,98 +70,171 @@ via the add button on the toolbar.
...
+const asRedCard = asToken({
+  withDesign({ Wrapper: withRedBorder }),
+  asToken.meta.term('Color')('Red'),
+});
+const ColoredCardsFC = withDesign({
+  BlueCard: asToken(startWith(Card), asDefaultCard, asBlueCard),
+  RedCard: asToken(startWith(Card), asDefaultCard, asRedCard),
+})(FlowContainer);
+```
+The above will cause the component selector to display a  "Color" filter
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/README.md:114]:
```suggestion
The above will cause the component selector to display a "Color" filter
```

File: packages/bodiless-layouts/README.md [hunk 10]
```diff
@@ -70,98 +70,171 @@ via the add button on the toolbar.
...
+const asBlueCard = asToken({
+  withDesign({ Wrapper: withBlueBorder }),
+  asToken.meta.term('Color')('Blue'),
+  asToken.meta.term('Design System')('JnJ'),
+});
+const FCWithBlacklistedCategory = asToken(
+  withDesign({
+    BlueCard: asToken(startWith(Card), asDefaultCard, asBlueCard),
+  }),
+  addProps({ blacklistCateogries: ['Design System' ] }),
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/README.md:153]:
```suggestion
  addProps({ blacklistCateogries: ['Design System'] }),
```

File: packages/bodiless-layouts/README.md [hunk 11]
```diff
@@ -204,22 +277,64 @@ corresponds and a class to used.
...
+The component selector displays a preview of each component to help the
+content editor understand what she is selecting.  By default, this is
+simply a scaled version of the original component. In some cases, it may
+be desirable to render a custom preview (for example, the actual rich text
+component is just an empty box, so instead we render its format bar as
+a preview).
+
+To facilitate custom previews, the flow container provides a React context
+informing each component whether it is being rendered in the component
+selector or on the actua page.  A component can use this to change the
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-layouts/README.md:314]:
```suggestion
selector or on the actual page.  A component can use this to change the
```

File: examples/test-site/src/data/pages/fc-basics/Basic.tsx [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+// These are metadata categories which should be ignored in
+// the flow container filters.
+const blacklistCategories = ['Category', 'Attribute', 'Component'];
+
+// This is the base component which will be included in all variations.
+const base = {
+  Box: asToken(startWith(Div) as HOC, asBox),
+};
+
+// Available colors
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/Basic.tsx:24]: Let's discuss.  I think variation here is acceptable.

File: examples/test-site/src/data/pages/fc-basics/Basic.tsx [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  withTailwindWidthConstraints,
+)(tailwindConfig);
+
+const asBasicFlowContainer = asToken(
+  // withDesign(basicDesign),
+  withAllTitlesFromTerms({ blacklistCategories }),
+  withDesign(variations) as HOC,
+  addProps({ blacklistCategories }),
+  addProps({ mandatoryCategories: ['Color'] }),
+  // withWidthConstraints('w-100 w-1/2 sm:w-1/2 sm:w-full lg:w-1/2 lg:w-full'),
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/Basic.tsx:96]: see above

File: examples/test-site/src/data/pages/fc-basics/ChameleonBox.tsx [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+});
+
+export const useSwapOverrides = () => ({
+  gropuLabel: 'Fill',
+});
+
+const ChameleonBox = asToken(
+  asBox,
+  asBodilessChameleon('chameleon', undefined, useToggleOverrides),
+  // asBodilessChameleon('chameleon', undefined, useSwapOverrides),
```
[the remark this answers is not in the document]
Comment (wodenx) [reply examples/test-site/src/data/pages/fc-basics/ChameleonBox.tsx:19]: see above

**Review by Guilherme-Almeida-Zeni — DISMISSED** [review]

**Comment by hvanyo** [comment]
@wodenx 
Issue 1: No active chameleon context.
STR: 
Add a component to /fc-basics/
Click on component underneath

ER:
no error and I can fill color

AR
No active chamelon context

**Comment by hvanyo** [comment]
@wodenx  what are you two test cases to test this?  I didn't see one without a title and they all had a color (mandatory field)

**Review by marcopagliarulo — APPROVED** [review]

File: examples/test-site/src/data/pages/fc-basics/ChameleonBox.tsx [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  groupLabel: 'Fill',
+});
+
+const ChameleonBox = asToken(
+  asBox,
+  asBodilessChameleon('chameleon', undefined, useToggleOverrides),
+  // asBodilessChameleon('chameleon', undefined, useSwapOverrides),
+  withBlueBorder,
+  withDesign({
+    Teal: asToken(asTeal, { title: 'Color Tesl' }),
```
Comment (hvanyo) [inline examples/test-site/src/data/pages/fc-basics/ChameleonBox.tsx:25]:
```suggestion
    Teal: asToken(asTeal, { title: 'Color Teal' }),
```

File: examples/test-site/src/data/pages/fc-basics/Box.tsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+export const asBox = asToken(
+  addClasses('w-full h-36'),
+  asToken.meta.term('Component')('Box'),
+  asToken.meta.term('Attribute')('Styles'),
+  asToken.meta.term('Type')('Box'),
+);
+
+export const withBlueBorder = asToken(
+  addClasses('border-blue-800 border-8'),
+  asToken.meta.term('Color')('Blue=Border'),
```
Comment (hvanyo) [inline examples/test-site/src/data/pages/fc-basics/Box.tsx:12]:
```suggestion
  asToken.meta.term('Color')('Blue-Border'),
```

**Comment by hvanyo** [comment]
@wodenx   this seems odd was this your intention?
STR

1. Click add on /fc-basis and select no components

ER:
I should see all three categories.

AR:
I only see colors until I choose a component

https://www.screencast.com/t/sVL7ccUzP5

**Comment by hvanyo** [comment]
@wodenx  I see test case now for component with mandatory color but I still don't find a component with no title

**Comment by wodenx** [comment]
> @wodenx I see test case now for component with mandatory color but I still don't find a component with no title

See https://github.com/johnsonandjohnson/Bodiless-JS/pull/965/files#diff-921e5ec33e630691b57f48e458e20101175f008a0b30dd7fb1c7133262ee30a4R91

**Comment by wodenx** [comment]
> STR
> 
> 1. Click add on /fc-basis and select no components
> 
> ER:
> I should see all three categories.
> 
> AR:
> I only see colors until I choose a component
This is expected.  Color is the only mandatory category, and there is now a component which has no categories.  FC only shows
- mandatory categories
- categories in which *all* the ciurrently filtered components have at least one term.
Once you select a color, the component with no metadata is filtered out, and all the currently filtered components have a term in the other categories.

**Review by hvanyo — APPROVED** [review]

**Review by Guilherme-Almeida-Zeni — APPROVED** [review]

**Pull request merged**